dag-json/src/codec.rs

```diff
@@ -0,0 +1,269 @@
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        self.visit_string(String::from(value))
+    }
+
+    #[inline]
+    fn visit_string<E>(self, value: String) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::String(value))
+    }
+    #[inline]
+    fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        self.visit_byte_buf(v.to_owned())
+    }
+
+    #[inline]
+    fn visit_byte_buf<E>(self, v: Vec<u8>) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Bytes(v))
+    }
+
+    #[inline]
+    fn visit_u64<E>(self, v: u64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Integer(v.into()))
+    }
+
+    #[inline]
+    fn visit_i64<E>(self, v: i64) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Integer(v.into()))
+    }
+
+    #[inline]
+    fn visit_i128<E>(self, v: i128) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Integer(v))
+    }
+
+    #[inline]
+    fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Bool(v))
+    }
+
+    #[inline]
+    fn visit_none<E>(self) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        self.visit_unit()
+    }
+
+    #[inline]
+    fn visit_unit<E>(self) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        Ok(Ipld::Null)
+    }
+
+    #[inline]
+    fn visit_seq<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
+    where
+        V: de::SeqAccess<'de>,
+    {
+        let mut vec: Vec<WrapperOwned> = Vec::new();
+
+        while let Some(elem) = visitor.next_element()? {
+            vec.push(elem);
+        }
+
+        let unwrapped = vec.into_iter().map(|WrapperOwned(ipld)| ipld).collect();
+        Ok(Ipld::List(unwrapped))
+    }
+
+    #[inline]
+    fn visit_map<V>(self, mut visitor: V) -> Result<Self::Value, V::Error>
+    where
+        V: de::MapAccess<'de>,
+    {
+        let mut values: Vec<(String, WrapperOwned)> = Vec::new();
+
+        while let Some((key, value)) = visitor.next_entry()? {
+            values.push((key, value));
+        }
+
+        // JSON Object represents IPLD Link if it is `{ "/": "...." }` therefor
+        // we valiadet if that is the case here.
+        if let Some((key, WrapperOwned(Ipld::String(value)))) = values.first() {
+            if key == LINK_KEY && values.len() == 1 {
+                // TODO: Find out what is the expected behavior in cases where
+                // value is not a valid CID (or base64 endode string here). For
+                // now treat it as some other JSON Object.
+                let link = base64::decode(value).unwrap();
+                let cid = Cid::try_from(link).unwrap();
+                return Ok(Ipld::Link(cid));
```

Comment:
shouldn't panic, but probably requires quite a bit more code. anyone familiar with serde? @Gozala

Reply:
If the `&'de str` could be acquired for this value then it would be a matter of returning https://docs.rs/serde/1.0.106/serde/de/trait.Error.html#method.invalid_value or `invalid_length` for these.

In the mean time the following:

```rust
                let link = base64::decode(value)
                    .map_err(|e| serde::de::Error::custom(e))
                    .and_then(|link| Cid::try_from(link).map_err(|e| serde::de::Error::custom(e)))?;

                return Ok(Ipld::Link(link));
```

passes the test:

```rust
    #[test]
    fn decode_invalid() {
        let input = r#"{ "/": "invalidcid" }"#;

        decode(input.as_bytes()).unwrap_err();
    }
```

Reply:
Ah of course, when the `&'de str` is not available we can just have `'static` errors, this gives a nicer: `invalid base64 encoding in link "invalidcid": Invalid last symbol 100, offset 9. at line 1 column 21`:

```
                #[derive(Debug)]
                enum InvalidLink {
                    InvalidEncoding(String, base64::DecodeError),
                    InvalidCid(String, libipld_base::cid::Error),
                }

                impl std::fmt::Display for InvalidLink {
                    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
                        match self {
                            InvalidLink::InvalidEncoding(s, e) => {
                                write!(fmt, "invalid base64 encoding in link {:?}: {}", s, e)
                            },
                            InvalidLink::InvalidCid(s, e) => {
                                write!(fmt, "invalid cid in link {:?}: {}", s, e)
                            }
                        }
                    }
                }

                impl std::error::Error for InvalidLink {}

                let raw_cid = match base64::decode(value.as_str()) {
                    Ok(bytes) => bytes,
                    Err(e) => {
                        if let (_, WrapperOwned(Ipld::String(value))) = values.swap_remove(0) {
                            return Err(serde::de::Error::custom(InvalidLink::InvalidEncoding(value, e)));
                        } else {
                            unreachable!("IPLD variant already checked");
                        }
                    }
                };

                let cid = match Cid::try_from(raw_cid) {
                    Ok(cid) => cid,
                    Err(e) => {
                        if let (_, WrapperOwned(Ipld::String(value))) = values.swap_remove(0) {
                            return Err(serde::de::Error::custom(InvalidLink::InvalidCid(value, e)));
                        } else {
                            unreachable!("IPLD variant already checked");
                        }
                    }
                };

                return Ok(Ipld::Link(cid));
```

Reply:
PR'd: https://github.com/dvc94ch/rust-ipld/pull/1